opm/simulators/linalg/ISTLSolverEbos.hpp

```diff
@@ -785,103 +375,74 @@ DenseMatrix transposeDenseMatrix(const DenseMatrix& M)
-            // For some reason simulator_.model().elementMapper() is not initialized at this stage.
-            // Hence const auto& elemMapper = simulator_.model().elementMapper(); does not work.
-            // Set it up manually
-            using ElementMapper =
-                Dune::MultipleCodimMultipleGeomTypeMapper<GridView>;
-            ElementMapper elemMapper(gridView, Dune::mcmgElementLayout());
-            typedef typename Matrix::size_type size_type;
-            size_type numCells = grid.size( 0 );
-            noGhostMat_.reset(new Matrix(numCells, numCells, Matrix::random));
-
-            std::vector<std::set<size_type>> pattern;
-            pattern.resize(numCells);
 
-            auto elemIt = gridView.template begin<0>();
-            const auto& elemEndIt = gridView.template end<0>();
-
-            //Loop over cells
-            for (; elemIt != elemEndIt; ++elemIt)
-            {
-                const auto& elem = *elemIt;
-                size_type idx = elemMapper.index(elem);
-                pattern[idx].insert(idx);
-
-                // Add well non-zero connections
-                for (auto wc = wellConnectionsGraph_[idx].begin(); wc!=wellConnectionsGraph_[idx].end(); ++wc)
-                    pattern[idx].insert(*wc);
-
-                // Add just a single element to ghost rows
-                if (elem.partitionType() != Dune::InteriorEntity)
-                {
-                    noGhostMat_->setrowsize(idx, pattern[idx].size());
-                }
-                else {
-                    auto isend = gridView.iend(elem);
-                    for (auto is = gridView.ibegin(elem); is!=isend; ++is)
-                    {
-                        //check if face has neighbor
-                        if (is->neighbor())
-                        {
-                            size_type nid = elemMapper.index(is->outside());
-                            pattern[idx].insert(nid);
-                        }
-                    }
-                    noGhostMat_->setrowsize(idx, pattern[idx].size());
+        /// Return true if we should (re)create the whole solver,
+        /// instead of just calling update() on the preconditioner.
+        bool shouldCreateSolver() const
+        {
+            // Decide if we should recreate the solver or just do
+            // a minimal preconditioner update.
+            if (!flexibleSolver_) {
+                return true;
+            }
+            if (this->parameters_.cpr_reuse_setup_ == 0) {
+                // Always recreate solver.
+                return true;
+            }
+            if (this->parameters_.cpr_reuse_setup_ == 1) {
+                // Recreate solver on the first iteration of every timestep.
+                const int newton_iteration = this->simulator_.model().newtonMethod().numIterations();
+                if (newton_iteration == 0) {
+                    return true;
                 }
```

Comment:
Writing it this way will trigger the [`assert`](https://github.com/OPM/opm-simulators/blob/30a4fe55b060dbf940565dc4fed22a15d318cdad/opm/simulators/linalg/ISTLSolverEbos.hpp#L407) at the end of the function if `newton_iteration > 0`.  You really need an unconditional `return` here&mdash;or remove the `assert`&mdash;if you don't want that to happen.